<launch>
	<arg name="sim" default="false" />
	<include file="$(find kcctslam)/launch/sensor.launch">
		<arg name="sim" value="$(arg sim)" />
	</include>

   <arg name="map_file0" default="$(find kcctslam)/config/map/sim1.yaml" />
   <arg name="map_file1" default="$(find kcctslam)/config/map/g201113-2.yaml" />
   <arg name="map_file2" default="$(find kcctslam)/config/map/g203.yaml" />
   <arg name="map_file3" default="$(find kcctslam)/config/map/g202.yaml" />
   <arg name="map_file4" default="$(find kcctslam)/config/map/g202.yaml" />

<!--    <node pkg="tf" type="static_transform_publisher" name="rs_odom_transform" args="0 0 0 0 0 0 rs_odom rs_odom_frame 100"/>-->
<!--	<node pkg="tf" type="static_transform_publisher" name="stf_hokuyo3d_laser" args="0.225 -0.255 0.0 0.0 0.0 0.0 camera_link 2d_blue 100"/>
	<node pkg="tf" type="static_transform_publisher" name="stf_hokuyo3d_laser2" args="-0.485 0.255 0.0 3.14159 0.0 0.0 camera_link 2d_orange 100"/>-->	
	<!--<node pkg="tf" type="static_transform_publisher" name="map_transform" args="0 0 0 0 0 0 map base_link 100"/>-->
<!--	<node pkg="tf" type="static_transform_publisher" name="base_transform" args="0 0 0 0 0 0 base_link camera_link 100"/>-->
<!--	<node pkg="tf" type="static_transform_publisher" name="rs_base_link_transform" args="-0.2 0 0 0 0 0 rs_link rs_camera_link 100"/>-->
	
	<rosparam param="footprint" ns="/move_base/local_costmap">[[0.05,-0.2], [0.05,0.2], [-0.3,0.2], [-0.3,-0.2]]</rosparam>
	<rosparam param="footprint" ns="/move_base/global_costmap">[[0.05,-0.2], [0.05,0.2], [-0.3,0.2], [-0.3,-0.2]]</rosparam>


	<node respawn="true" name="map_server0" pkg="map_server" type="map_server" args="$(arg map_file0)"><remap from="/map" to="/map0" /></node>
	<node respawn="true" name="map_server1" pkg="map_server" type="map_server" args="$(arg map_file1)"><remap from="/map" to="/map1" /><remap from="/static_map" to="/static_map1" /></node>
	<!--<node respawn="true" name="map_server2" pkg="map_server" type="map_server" args="$(arg map_file2)"><remap from="/map" to="/map2" /><remap from="/static_map" to="/static_map2" /></node>
	<node respawn="true" name="map_server3" pkg="map_server" type="map_server" args="$(arg map_file3)"><remap from="/map" to="/map3" /><remap from="/static_map" to="/static_map3" /></node>
	<node respawn="true" name="map_server4" pkg="map_server" type="map_server" args="$(arg map_file4)"><remap from="/map" to="/map4" /><remap from="/static_map" to="/static_map4" /></node>
-->
	<node pkg="kcctslam" type="map_selecter" name="map_selecter" output="screen"/>
	<!--
	<node name="joy_node" pkg="joy" type="joy_node" />
	<node name="basic_twist_publisher" pkg="basic_lecture" type="basic_twist_publisher">
	</node>
	value="57600"
	-->
	
	
	
	
	
	<node name="amcl" pkg="amcl" type="amcl">
		<remap from="/scan" to="/scan1" />	
		<!--<param name="odom_alpha1" type="double" value="7" />
		<param name="odom_alpha2" type="double" value="7" />
		<param name="odom_alpha3" type="double" value="3.5" />
		<param name="odom_alpha4" type="double" value="3.5" />-->
		
		<param name="odom_alpha1" type="double" value="10" />
		<param name="odom_alpha2" type="double" value="7" />
		<param name="odom_alpha3" type="double" value="3.5" />
		<param name="odom_alpha4" type="double" value="2.4" />
		
		
		<!--ＮＥＣで良かったパラメータ-->
		<!--<param name="odom_alpha1" type="double" value="17" />
		<param name="odom_alpha2" type="double" value="15" />
		<param name="odom_alpha3" type="double" value="8.5" />
		<param name="odom_alpha4" type="double" value="6.4" />
		-->
		<param name="laser_max_beams" type="int" value="150" />
		<param name="min_particles" type="int" value="300" />
		<param name="max_particles" type="int" value="1000" />
		<param name="update_min_d" type="double" value="0.05" />
		<param name="update_min_a" type="double" value="0.08" />
		<param name="use_map_topic" type="bool" value="false" />
	</node>

	<node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen">
		<rosparam file="$(find kcctslam)/config/nav_param/move_base.yaml" command="load" />
		<rosparam file="$(find kcctslam)/config/nav_param/costmap_common.yaml" command="load" ns="global_costmap" />
		<rosparam file="$(find kcctslam)/config/nav_param/costmap_common.yaml" command="load" ns="local_costmap" />
		<rosparam file="$(find kcctslam)/config/nav_param/costmap_local.yaml" command="load" />
		<rosparam file="$(find kcctslam)/config/nav_param/costmap_global.yaml" command="load" />
		<rosparam file="$(find kcctslam)/config/nav_param/planner_local_yvt.yaml" command="load" />
		<rosparam file="$(find kcctslam)/config/nav_param/planner_global.yaml" command="load" />
	</node>
	
	<!-- To run python scripts, chmod 755 to file is required (in this case, chmod 755 laser2pc.py). -->
	<group unless="$(arg sim)">
		<node pkg="kcctslam" name="laser2pc" type="laser2pc.py" output="screen">
		</node>


		<!-- 近傍点探索に使うパラメータと結果が入る変数
		radius: 半径 r [m].
		max_nn: 最近傍のお隣さんを何点見つかったら探索を打ち切るか。(nn = nearest neighbor)
			0にすると打ち切らない
		-->	
		
		<node pkg="kcctslam" name="pcl_handler" type="pcl_handler" >
			<param name="max_nn" value="20"/>
			<param name="radius" value="10.0"/>
			<param name="x_max" value="5.0"/>
			<param name="x_min" value="-0.5"/>
			<param name="y_max" value="0.4"/>
			<param name="y_min" value="-0.4"/>
			<param name="loop_rate" value="10"/>		
		</node>
	</group>
	
	<node pkg="turtlesim" type="turtle_teleop_key" name="turtle_teleop_key" output="screen"/>
	
	<!--<node name="waypoint" pkg="nakanoshima" type="wpmove.py" output="screen"/>-->
	<!--<node pkg="kcctslam" type="robot_model_node" name="robot_model_node" output="screen"/>-->
	
	
	<node pkg="rviz" type="rviz" args="-d '$(find kcctslam)/config/rviz/navigation_yvtx004.rviz'" name="visualizer" respawn="true" />
	<!--<node name="odom_pub" pkg="kcctslam" type="odom_pub"/>-->
	<!--<node name="tf_pub" pkg="kcctslam" type="tf_pub"/>-->
	<!--<include file="$(find ublox_gps)/launch/ublox_zed-f9p.launch"/>-->
	
</launch>
